src/core/qgsmaplayerlegend.cpp

```diff
@@ -198,11 +198,33 @@ QList<QgsLayerTreeModelLegendNode*> QgsDefaultVectorLayerLegend::createLayerTree
     nodes.append( new QgsSimpleLegendNode( nodeLayer, r->legendClassificationAttribute() ) );
   }
 
+  // we have varying icon sizes, and we want icon to be centered and
+  // text to be left aligned, so we have to compute the max size of icons
+  QList<QgsSymbolV2LegendNode*> symbolNodes;
+  int widthMax = 0;
   foreach ( const QgsLegendSymbolItemV2& i, r->legendSymbolItemsV2() )
   {
-    nodes.append( new QgsSymbolV2LegendNode( nodeLayer, i ) );
+    QgsSymbolV2LegendNode * n = new QgsSymbolV2LegendNode( nodeLayer, i );
+    nodes.append( n );
+    if ( i.symbol() )
+    {
+      symbolNodes.append( n );
+      n->setCrop( true );
+      QPixmap pix( n->data( Qt::DecorationRole ).value<QPixmap>() );
+      widthMax = qMax( pix.width(), widthMax );
+      n->setCrop( false );
+      n->setIconSize( pix.size() );
+    }
+  }
+
+  foreach ( QgsSymbolV2LegendNode* n, symbolNodes )
+  {
+    Q_ASSERT( widthMax > 0 );
+    n->setIconSize( QSize( widthMax + 2, n->iconSize().rheight() + 2 ) );
+    n->setCrop( false );
```

Comment:
If I understand the code correctly, it temporarily turns on cropping to get the minimum size of each symbol, calculates the max width and finally sets the icon size for all icons to match the width. After initialization, cropping is always disabled.

- could you remove crop()/setCrop() from the API and just have a call minimumIconSize() to get that cropped icon size? I would prefer to keep such code away from data() implementation as it is unclear the cropping is enabled just temporarily.
- why +2 for icon sizes? (better always comment why magic numbers are used)
- how about rule-based renderer? it shows legend in a tree, so I guess there should be different max width for each sub-tree (otherwise one wide icon somewhere in the tree will make everything wide)